Add PInvoke/ExactSpelling tests

jiangzeng opened the pull request.

jiangzeng01 committed:
Add PInvoke/ExactSpelling tests

Refactor tests to fit with the rest of the Interop tests.

Fix up test to cleanly run.

Change CMakeLists.txt to match the rest of the tests.

Include Interop.cmake in CMakeLists.txt

Remove Service.

AaronRobinsonMSFT reviewed (dismissed):
Once all green - LGTM

AaronRobinsonMSFT commented on tests/src/Interop/PInvoke/ExactSpelling/CMakeLists.txt:
```diff
@@ -1,8 +1,13 @@
-cmake_minimum_required (VERSION 2.6) 
-project (ExactSpellingNative) 
-include ("${CLR_INTEROP_TEST_ROOT}/Interop.cmake") 
```
I would prefer if you could keep this line. See the `readme.md` at the root of the Interop test dir for details.

jkoritzinsky committed:
On x86 enable stdcall mangling irrespective of ExactSpelling and account for the charset suffix when ExactSpelling = false.

Change variable name.

Clean up the FindEntryPoint. The logic flow now matches CoreRT + CoreCLR specific features (ordinals and stdcall mangling).

PR Feedback.

Fix format specifier.

Add back probing null check.

Fix offset calculation for stdcall mangling.

Probe the stdcall-mangled versions of the original entry-point names when ExactSpelling isn't set.

Cleanup.

jkoritzinsky commented:
@AaronRobinsonMSFT I found some bugs with our proc address resolution + x86 stdcall mangling. Can you take another review pass when you get a chance?

AaronRobinsonMSFT commented on src/vm/method.cpp:
```diff
@@ -4994,24 +5000,29 @@ LPVOID NDirectMethodDesc::FindEntryPoint(HINSTANCE hMod) const
     DWORD nbytes = (DWORD)(strlen(GetEntrypointName()) + 1);
```
Instead of `GetEntrypointName()` we should use the `funcName` local or remove the variable entirely.

AaronRobinsonMSFT commented on src/vm/method.cpp:
```diff
@@ -4959,84 +4959,80 @@ LPVOID NDirectMethodDesc::FindEntryPoint(HINSTANCE hMod) const
-        if (pFuncW != NULL)
+        FARPROC pProbedFunc = GetProcAddress(hMod, szProbedEntrypointName);
+
+        if (pProbedFunc != NULL)
```
This check isn't needed. Just assign directly to `pFunc` instead of creating a local.

jkoritzinsky replied:
This check actually is needed. In the case that the user-provided name resolves to a function and is marked Unicode, but the user-provided name + 'W' doesn't resolve.

AaronRobinsonMSFT commented on src/vm/method.cpp:
```diff
@@ -4959,84 +4959,80 @@ LPVOID NDirectMethodDesc::FindEntryPoint(HINSTANCE hMod) const
         }
-#endif // _TARGET_X86_
+
+        szProbedEntrypointName[-1] = '_';
```
Logic is sound, but usage requires propagation of pointer arithmetic knowledge. Instead update the pointer to be correct and just use it naturally (i.e. `szProbedEntrypointName = szProbedEntrypointName - 1;`)

jkoritzinsky commented:
@dotnet-bot test this please

AaronRobinsonMSFT commented on src/vm/method.cpp:
```diff
@@ -4959,84 +5003,56 @@ LPVOID NDirectMethodDesc::FindEntryPoint(HINSTANCE hMod) const
     }
 #endif
 
     // Just look for the unmangled name.  If it is unicode fcn, we are going
```
Comment is stale.

AaronRobinsonMSFT commented on src/vm/method.cpp:
```diff
@@ -4959,84 +5003,56 @@ LPVOID NDirectMethodDesc::FindEntryPoint(HINSTANCE hMod) const
     // Allocate a single character before the start of this string to enable quickly
     // prepending a '_' character if we look for a stdcall entrypoint later on.
-    LPSTR szAnsiEntrypointName = ((LPSTR)_alloca(dstbufsize + 1)) + 1;
+    LPSTR szProbedEntrypointName = ((LPSTR)_alloca(dstbufsize + 1)) + 1;
```
Do we still need the additional `+1`s here? Seems like you handled that above.

AaronRobinsonMSFT reviewed (approved):
You might want to clean up these commits locally and force update your branch. Having product changes in a single commit and test changes in another helps if we need to port product changes and not test changes.
